src/jit.c

```diff
@@ -0,0 +1,1557 @@
+    pthread_jit_write_protect_np(true);
+#endif
+}
+#endif
+
+static inline void emit_jump_target_offset(struct jit_state *state,
+                                           uint32_t jump_loc,
+                                           uint32_t jump_state_offset)
+{
+    struct jump *jump = &state->jumps[state->n_jumps++];
+    jump->offset_loc = jump_loc;
+    jump->target_offset = jump_state_offset;
+}
+
+/* The REX prefix and ModRM byte are emitted.
+ * The MR encoding is utilized when a choice is available. The 'src' is often
+ * used as an opcode extension.
+ */
+static inline void emit_alu32(struct jit_state *state, int op, int src, int dst)
+{
+#if defined(__x86_64__)
+    emit_basic_rex(state, 0, src, dst);
+    emit1(state, op);
+    emit_modrm_reg2reg(state, src, dst);
+#elif defined(__aarch64__)
+    switch (op) {
+    case 1: /* ADD */
+        emit_addsub_register(state, false, AS_ADD, dst, dst, src);
+        break;
+    case 0x29: /* SUB */
+        emit_addsub_register(state, false, AS_SUB, dst, dst, src);
+        break;
+    case 0x31: /* XOR */
+        emit_logical_register(state, false, LOG_EOR, dst, dst, src);
+        break;
+    case 9: /* OR */
+        emit_logical_register(state, false, LOG_ORR, dst, dst, src);
+        break;
+    case 0x21: /* AND */
+        emit_logical_register(state, false, LOG_AND, dst, dst, src);
+        break;
+    case 0xd3:
+        if (src == 4) /* SLL */
+            emit_dataproc_2source(state, false, DP2_LSLV, dst, dst, R8);
+        else if (src == 5) /* SRL */
+            emit_dataproc_2source(state, false, DP2_LSRV, dst, dst, R8);
+        else if (src == 7) /* SRA */
+            emit_dataproc_2source(state, false, DP2_ASRV, dst, dst, R8);
+        break;
+    default:
+        __UNREACHABLE;
+        break;
+    }
+#endif
+}
+
+/* REX prefix, ModRM byte, and 32-bit immediate */
+static inline void emit_alu32_imm32(struct jit_state *state,
+                                    int op UNUSED,
+                                    int src,
+                                    int dst,
+                                    int32_t imm)
+{
+#if defined(__x86_64__)
+    emit_alu32(state, op, src, dst);
+    emit4(state, imm);
+#elif defined(__aarch64__)
+    switch (src) {
+    case 0:
+        emit_load_imm(state, R10, imm);
+        emit_addsub_register(state, false, AS_ADD, dst, dst, R10);
+        break;
+    case 1:
+        emit_load_imm(state, R10, imm);
+        emit_logical_register(state, false, LOG_ORR, dst, dst, R10);
+        break;
+    case 4:
+        emit_load_imm(state, R10, imm);
+        emit_logical_register(state, false, LOG_AND, dst, dst, R10);
+        break;
+    case 6:
+        emit_load_imm(state, R10, imm);
+        emit_logical_register(state, false, LOG_EOR, dst, src, R10);
+        break;
+    default:
+        __UNREACHABLE;
+        break;
+    }
+#endif
+}
+
+/* REX prefix, ModRM byte, and 8-bit immediate */
```

Comment:
This comment should only make sense to x64 backend. Move into the function body.